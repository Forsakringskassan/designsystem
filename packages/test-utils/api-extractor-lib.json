{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
