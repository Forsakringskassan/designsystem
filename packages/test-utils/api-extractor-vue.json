{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
