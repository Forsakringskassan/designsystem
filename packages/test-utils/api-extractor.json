{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
