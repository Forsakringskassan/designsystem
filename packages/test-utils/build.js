const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
