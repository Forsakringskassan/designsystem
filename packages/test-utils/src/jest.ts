/* imported for side-effects only */
import "./matchers";

export * from "./lib";
