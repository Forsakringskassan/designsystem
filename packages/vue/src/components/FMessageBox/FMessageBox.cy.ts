import { defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import FMessageBox from "./FMessageBox.vue";

function getShortTemplate(type: string): string {
    return /* HTML */ `
        <density-wrapper>
            <f-message-box type="${type}" layout="short">
                Kort text
            </f-message-box>
        </density-wrapper>
    `;
}

function getDefaultTemplate(type: string): string {
    return /* HTML */ `
        <density-wrapper>
            <f-message-box type="${type}">
                <template #default="{ headingSlotClass }">
                    <h2 :class="headingSlotClass">Rubrik</h2>
                    <p>Br√∂dtext</p>
                </template>
            </f-message-box>
        </density-wrapper>
    `;
}

describe("default layout", () => {
    const DensityComponent = defineComponent({
        template: getDefaultTemplate("info"),
        components: {
            DensityWrapper,
            FMessageBox,
        },
    });

    it(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.get(".message-box").should("be.visible");
        cy.toMatchScreenshot();
    });
});

describe("short layout", () => {
    const DensityComponent = defineComponent({
        template: getShortTemplate("info"),
        components: {
            DensityWrapper,
            FMessageBox,
        },
    });

    it(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.get(".message-box").should("be.visible");
        cy.toMatchScreenshot();
    });
});

it("should have approved design", () => {
    const ScreenshotComponent = defineComponent({
        template: /* HTML */ `
            <div class="row">
                <div class="col col--sm-6">
                    ${getDefaultTemplate("success")}
                    ${getDefaultTemplate("warning")}
                    ${getDefaultTemplate("error")} ${getDefaultTemplate("info")}
                </div>
                <div class="col col--sm-6">
                    ${getShortTemplate("success")}
                    ${getShortTemplate("warning")} ${getShortTemplate("error")}
                    ${getShortTemplate("info")}
                </div>
            </div>
        `,
        components: {
            FMessageBox,
        },
    });

    cy.mount(ScreenshotComponent);
    cy.get(".message-box").should("be.visible");
    cy.toMatchScreenshot();
});
