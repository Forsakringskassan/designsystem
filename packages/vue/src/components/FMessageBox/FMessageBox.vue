<script lang="ts">
import { type PropType, defineComponent } from "vue";
import { TranslationService } from "@fkui/logic";
import FIcon from "../FIcon/FIcon.vue";
import { IFlex, IFlexItem } from "../../internal-components/IFlex";

const iconClasses = {
    success: {
        symbol: "circle",
        sign: "success",
        screenReaderContextKey: "fkui.message-box.sr-context.success",
        screenReaderContextDefault: "Meddelande",
    },
    warning: {
        symbol: "circle",
        sign: "alert",
        screenReaderContextKey: "fkui.message-box.sr-context.warning",
        screenReaderContextDefault: "Varningsmeddelande",
    },
    error: {
        symbol: "triangle",
        sign: "alert",
        screenReaderContextKey: "fkui.message-box.sr-context.error",
        screenReaderContextDefault: "Felmeddelande",
    },
    info: {
        symbol: "circle",
        sign: "i",
        screenReaderContextKey: "fkui.message-box.sr-context.info",
        screenReaderContextDefault: "Informationsmeddelande",
    },
} as Record<string, Record<string, string>>;

export default defineComponent({
    name: "FMessageBox",
    components: {
        FIcon,
        IFlex,
        IFlexItem,
    },
    props: {
        /**
         * Type of message-box. 'success', 'error', 'warning' and 'info' is valid.
         * */
        type: {
            type: String,
            required: true,
            validator(value: string) {
                // The value must match one of these strings
                return ["success", "warning", "error", "info"].includes(value);
            },
        },
        /**
         * If message-box should be a banner.
         * If prop is not used message-box will have default styling.
         */
        banner: {
            type: Boolean,
            required: false,
        },
        /**
         * It is important to provide a context for a screenreader, similar to a modal that initially outputs it is a dialog.
         * Type-specific screenreader context is output by default in this component.
         * By setting this property to false no context will be output.
         * Note that by doing this, it is the applications responsiblity to provide a clear context.
         */
        provideScreenReaderContext: {
            type: Boolean,
            required: false,
            default: true,
        },
        /**
         * Select standard or short layout.
         *
         * - 'normal' - Use normal layout when need for heading and/or longer text/multiple sentences
         * - 'short' - Use short layout when only need for shorter text
         */
        layout: {
            type: String as PropType<"standard" | "short">,
            required: false,
            default: "standard",
            validator(value: string) {
                return ["standard", "short"].includes(value);
            },
        },
    },
    data() {
        return {
            headingClass: ["message-box__heading"],
        };
    },
    computed: {
        messageBoxType(): string {
            if (this.layout === "short") {
                return `message-box--${this.type}-short`;
            } else {
                return `message-box--${this.type}`;
            }
        },
        bannerType(): string {
            return this.banner ? `message-box--banner` : "";
        },
        classType(): string {
            return `icon__${this.type}`;
        },
        classIcon(): string {
            return iconClasses[this.type].symbol === "circle" && iconClasses[this.type].sign === "alert"
                ? `icon__exclamation`
                : "";
        },
        stackTypeClass(): string {
            return `icon-stack--${this.type}`;
        },
        symbol(): string {
            return iconClasses[this.type].symbol;
        },
        sign(): string {
            return iconClasses[this.type].sign;
        },
    },
    methods: {
        screenReaderContext() {
            return TranslationService.provider.translate(
                iconClasses[this.type].screenReaderContextKey,
                iconClasses[this.type].screenReaderContextDefault,
            );
        },
    },
});
</script>

<template>
    <div class="message-box" :class="[messageBoxType, bannerType]">
        <span v-if="provideScreenReaderContext" class="sr-only">{{ screenReaderContext() }}</span>

        <i-flex gap="2x">
            <i-flex-item v-if="layout === 'short'" class="message-box__icon" shrink align="center">
                <span class="icon-stack" :class="stackTypeClass">
                    <f-icon :class="classType" :name="symbol"></f-icon>
                    <f-icon :class="classIcon" :name="sign"></f-icon>
                </span>
            </i-flex-item>
            <i-flex-item class="message-box__content" grow align="center">
                <!--
                @slot Slot for displaying content. Slot styling is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ headingSlotClass }"`.
                @binding {string[]} headingSlotClass CSS class to use for the heading element. Only set for normal layout (avoid heading otherwise). Use with `:class="headingSlotClass"`.
                -->
                <slot v-bind="layout === 'short' ? {} : { headingSlotClass: headingClass }"></slot>
            </i-flex-item>
        </i-flex>
    </div>
</template>
