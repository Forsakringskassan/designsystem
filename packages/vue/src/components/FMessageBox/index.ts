export { default as FMessageBox } from "./FMessageBox.vue";
