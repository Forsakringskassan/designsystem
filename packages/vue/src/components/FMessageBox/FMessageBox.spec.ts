import { VueWrapper, mount } from "@vue/test-utils";
import FMessageBox from "./FMessageBox.vue";
import "html-validate/jest";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FMessageBox, {
        attrs: { ...attrs },
        props: { type: "", ...props },
        slots: { ...slots },
        global: {
            stubs: ["FIcon"],
        },
    });
}

describe("FMessageBox", () => {
    it.each`
        type
        ${"info"}
        ${"error"}
        ${"warning"}
        ${"success"}
    `("should match snapshot with correct type", ({ type }) => {
        const wrapper = createWrapper({ props: { type } });
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with content", () => {
        const wrapper = createWrapper({
            props: { type: "success" },
            slots: { default: ` <p>content</p> ` },
        });
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("html-validate", () => {
    it("should not report error when used correctly", () => {
        const markup = /* HTML */ `
            <f-message-box type="warning">
                <template v-slot="{ headingClass }">
                    <h2 :class="headingClass">Lorem ipsum</h2>
                    <p>dolor sit amet</p>
                </template>
            </f-message-box>
        `;
        expect(markup).toHTMLValidate();
    });

    it("should report error when obsolete heading slot is used", () => {
        const markup = /* HTML */ `
            <f-message-box type="warning">
                <template v-slot:heading></template>
            </f-message-box>
        `;
        expect(markup).not.toHTMLValidate();
    });

    it("should report error when type is missing", () => {
        const markup = /* HTML */ ` <f-message-box></f-message-box> `;
        expect(markup).not.toHTMLValidate();
    });

    it("should report error when type is invalid", () => {
        const markup = /* HTML */ `
            <f-message-box type="foobar"></f-message-box>
        `;
        expect(markup).not.toHTMLValidate();
    });
});
