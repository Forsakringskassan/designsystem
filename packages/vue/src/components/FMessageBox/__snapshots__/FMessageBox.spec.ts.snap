// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FMessageBox should match snapshot with content 1`] = `
<div
  class="message-box message-box--success"
>
  <span
    class="sr-only"
  >
    Meddelande
  </span>
  <div
    class="iflex iflex--gap-2x"
  >
    
    <!--v-if-->
    <div
      class="iflex__item iflex--align-center iflex--grow message-box__content"
    >
      
      
      <p>
        content
      </p>
      
      
    </div>
    
  </div>
</div>
`;

exports[`FMessageBox should match snapshot with correct type 1`] = `
<div
  class="message-box message-box--info"
>
  <span
    class="sr-only"
  >
    Informationsmeddelande
  </span>
  <div
    class="iflex iflex--gap-2x"
  >
    
    <!--v-if-->
    <div
      class="iflex__item iflex--align-center iflex--grow message-box__content"
    >
      
      
    </div>
    
  </div>
</div>
`;

exports[`FMessageBox should match snapshot with correct type 2`] = `
<div
  class="message-box message-box--error"
>
  <span
    class="sr-only"
  >
    Felmeddelande
  </span>
  <div
    class="iflex iflex--gap-2x"
  >
    
    <!--v-if-->
    <div
      class="iflex__item iflex--align-center iflex--grow message-box__content"
    >
      
      
    </div>
    
  </div>
</div>
`;

exports[`FMessageBox should match snapshot with correct type 3`] = `
<div
  class="message-box message-box--warning"
>
  <span
    class="sr-only"
  >
    Varningsmeddelande
  </span>
  <div
    class="iflex iflex--gap-2x"
  >
    
    <!--v-if-->
    <div
      class="iflex__item iflex--align-center iflex--grow message-box__content"
    >
      
      
    </div>
    
  </div>
</div>
`;

exports[`FMessageBox should match snapshot with correct type 4`] = `
<div
  class="message-box message-box--success"
>
  <span
    class="sr-only"
  >
    Meddelande
  </span>
  <div
    class="iflex iflex--gap-2x"
  >
    
    <!--v-if-->
    <div
      class="iflex__item iflex--align-center iflex--grow message-box__content"
    >
      
      
    </div>
    
  </div>
</div>
`;
